Implement donation levels for projects

erictshen opened the pull request:
Fixes #191 
revolv++

Lots of inspiration from http://www.lab305.com/news/2012/jul/19/django-inline-formset-underscore/
http://jacobian.org/writing/dynamic-form-generation/
http://kevindias.com/writing/django-class-based-views-multiple-inline-formsets/
https://docs.djangoproject.com/en/1.7/topics/forms/modelforms/#inline-formsets

erictshen commented:
![screen shot 2015-03-16 at 9 01 51 pm](https://cloud.githubusercontent.com/assets/2114482/6681197/c74119c0-cc1f-11e4-9468-587a60fccdec.png)
![screen shot 2015-03-16 at 9 02 22 pm](https://cloud.githubusercontent.com/assets/2114482/6681198/c741e6de-cc1f-11e4-86c9-1b519def93e0.png)
![screen shot 2015-03-16 at 9 02 49 pm](https://cloud.githubusercontent.com/assets/2114482/6681200/d612d47a-cc1f-11e4-9600-8f5376a96463.png)

noahsark769 commented on revolv/project/models.py:
```diff
@@ -395,8 +396,16 @@ def is_completed(self):
 
+    @property
+    def donation_levels(self):
+        return DonationLevel.objects.filter(project_id=self.id).values()
```
Woah, this works? project_id isn't a field of DonationLevel. At any rate, it would be better to just return self.donationlevel_set, correct?

noahsark769 commented on revolv/project/views.py:
```diff
@@ -62,15 +77,28 @@ def get_success_url(self):
     def form_valid(self, form):
-        project = self.get_object()
-        project.update_categories(form.cleaned_data['categories_select'])
+        context = self.get_context_data()
```
Can we implement like a `get_formset` method for this? It's fairly weird to be calling `get_context_data` in a POST request

erictshen replied:
Will look into this mid break (like Wednesday-ish). Unfortunately, there might not be a way to do this without calling get_context_data (though I can define a helper method)

noahsark769 replied:
I think helper method is the way to go
On Fri, Mar 20, 2015 at 11:35 PM Eric Shen <notifications@github.com> wrote:

> In revolv/project/views.py
> <https://github.com/calblueprint/revolv/pull/204#discussion_r26889714>:
>
> >      def form_valid(self, form):
> > -        project = self.get_object()
> > -        project.update_categories(form.cleaned_data['categories_select'])
> > +        context = self.get_context_data()
>
> Will look into this mid break (like Wednesday-ish). Unfortunately, there
> might not be a way to do this without calling get_context_data (though I
> can define a helper method)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/calblueprint/revolv/pull/204/files#r26889714>.
>

noahsark769 commented:
Just a couple of things (I commented on them). Also, can we at least format the perks in some sort of way other than the default python stringification? It's super weird to have that JSON like syntax in the template, even if it will be rewritten soon

noahsark769 commented:
@erictshen let me know when you can take a look at the get_context_data stuff

erictshen commented:
Added helper methods and renamed some of the objects and methods to make it more understandable. Let me know if there's anything else to change!

erictshen commented:
I see. Makes sense, and it works!

noahsark769 commented:
Okay wait, so now we have the exact same method in two classes. Can we extract that into a `DonationLevelFormsetMixin` (with appropriate docstring for what the one method it provides does).

Also, for categories display on the project page, you changed it to not have the JSON formatting right?

erictshen commented:
Updated pull request to use DonationLevelFormSetMixin

Also fixed project page to show categories not in JSON

![screen shot 2015-03-25 at 5 48 43 pm](https://cloud.githubusercontent.com/assets/2114482/6838507/d01cd486-d317-11e4-9660-359aeaebe74e.png)

noahsark769 commented on revolv/project/views.py:
```diff
@@ -15,7 +16,21 @@
+    Mixin that gets the ProjectDonationLeveLFormSet for a page, specifically
+    the Create Project and Update Project page.
+    """
+    # checks if the request is a POST, and populates the formset with current object as the instance
```
This should be a docstring of the method :)

noahsark769 commented:
Looks great!
